### DISCUSSION TITLE
How to apply fix its with the python API

### DISCUSSION THREAD

**Discussion by alcroito** [ast-grep/ast-grep#1004] [Q&A]
I've looked through the python and JS API, and it seems it's currently not possible to apply fix rules to nodes, but only find / match nodes, is that correct?

**Comment by HerringtonDarkholme** [comment]
yes, at the moment JS/Py API does not support fix.
The ideal way to apply fix is using https://github.com/Rich-Harris/magic-string in JavaScript. 
Fix in ast-grep is just string splice and concatenation.

**Reply by nwalters512** [reply]
It'd be good to update the docs to reflect this. Currently the API introduction says the following:

```
ast-grep allows you to craft complicated rules, but it is not easy to do arbitrary AST manipulation.

For example, you may struggle to:

- replace a list of nodes individually, based on their content
- replace a node conditionally, based on its content and surrounding nodes
- count the number or order of nodes that match a certain pattern
- compute the replacement string based on the matched nodes

To solve these problems, you can use ast-grep's programmatic API! You can freely inspect and change syntax trees in popular programming languages!
```

The "and change syntax trees" part got me so excited, but after spending a lot of time in the docs I finally ended up here to discover that one _cannot_ implement programmatic modifications. I'd of course love to see this implemented somehow, but in the meantime, it'd be nice if the docs reflected what's actually possible.

**Comment by HerringtonDarkholme** [comment]
Hi! ast-grep-py now has an experimental fix API https://github.com/ast-grep/ast-grep/blob/main/crates/pyo3/tests/test_fix.py

**Comment by HerringtonDarkholme** [comment]
I'm gonna close this and track the documentation in a new issue. Thanks!
